--- link.diff.inc.old	2014-05-28 09:29:15.189299280 +0200
+++ link.diff.inc	2014-05-28 09:30:27.133297848 +0200
@@ -11,11 +11,14 @@
 function link_field_diff_view($items, $context) {
   $diff_items = array();
   foreach ($items as $delta => $item) {
-    if ($item['url'] && $item['title']) {
+    if (isset($item['url']) && isset($item['title'])) {
       $diff_items[$delta] = $item['title'] . ' (' . $item['url'] . ')';
     }
+    elseif(isset($item['url'])) {
+      $diff_items[$delta] = $item['url'];
+    }
     else {
-      $diff_items[$delta] = $item['title'] . $item['url'];
+      $diff_items[$delta] = $item['title'];
     }
   }
   return $diff_items;
