--- state_flow.module.old	2014-04-18 11:59:52.045493445 +0200
+++ state_flow.module	2014-04-18 12:00:46.741492356 +0200
@@ -450,7 +450,11 @@
     // When a draft is saved and does not become the current revision, then
     // redirect the user to the revision saved. This hijacks the redirection by
     // drupal_goto().
-    $_GET['destination'] = 'node/' . $node->nid . '/revisions/' . $node->vid . '/view';
+    if ($node->type == 'panel') {
+      $_GET['destination'] = 'node/' . $node->nid . '/revisions/' . $node->vid . '/edit/view';
+    } else {
+      $_GET['destination'] = 'node/' . $node->nid . '/revisions/' . $node->vid . '/view';
+    }
 
     // Save the vid for the draft revision back to the node in case other modules need it.
     $node->new_draft_vid = $node->vid;
