--- state_flow.pages.inc.old	2014-06-10 12:00:43.259715800 +0200
+++ state_flow.pages.inc	2014-06-10 12:01:04.583715375 +0200
@@ -66,7 +66,7 @@
       $state_actions_str = implode(' | ', $state_actions);
       $item = array(
         l($vid, $path_view),
-        l(check_plain($state_node->title), $path_view),
+        l($state_node->title, $path_view),
         t($state->state),
         $state_date_str,
         theme('username', array('account' => $rev_author)),
