diff --git a/context_reaction_http_header.inc b/context_reaction_http_header.inc
index bef2584..4d2ba44 100644
--- a/context_reaction_http_header.inc
+++ b/context_reaction_http_header.inc
@@ -16,7 +16,9 @@ class context_reaction_http_header extends context_reaction {
     
     // Get set values from the context for these hooks  
     foreach ($items_from_hook as $header => $item) {
-      $items_from_hook[$header]['#default_value'] = $values['http_header'][$header];
+      if (isset($values['http_header'][$header])) {
+        $items_from_hook[$header]['#default_value'] = $values['http_header'][$header];
+      }
     }
     
     $items['http_header'] = array_merge($items['http_header'], $items_from_hook);
